### ISSUE TITLE
Implement postback deferral mechanism

### ISSUE THREAD

**Issue by tomasherceg** [riganti/dotvvm#105]
Currently, there is no way to display a confirmation, when the user clicks a button. We should introduce a mechanism that can defer and possibly cancel a postback. Something like this:

```
<dot:Button Click="{command: _parent.DeleteUser(Id)}" PostBack.Deferred="confirm" PostBack.DeferredOptions="Are you really want to delete this user?" />
```

The **confirm** is a name of a postback deferral provider, which can look like this for standard javascript confirmation (in Typescript):
```
class ConfirmPostBackDeferral {
    public initialize(options: any, viewModel: any, postBackCallback: () => any): void {
        if (confirm(options.toString())) {
            postBackCallback();
        }
    }
}
```

Or, if you need to display some nice looking modal dialog, it can look like this:
```
class CustomConfirmPostBackDeferral {
    public initialize(options: any, viewModel: any, postBackCallback: () => any): void {
        $("#confirm-dialog").show();
        $("#confirm-dialog .ok-button").unbind("click").bind("click", postBackCallback);
    }
}
```

Another usage can be found e.g. for search textboxes - we do not want to send the postback immediately when the user types a letter, however we need to delay the postback e.g. for 500ms.
```
class DelayPostBackDeferral {
    private counter: int = 0;
    public initialize(options: any, viewModel: any, postBackCallback: () => any): void {
          this.counter++;
          var currentCounter = this.counter;

          window.setTimeout(() => {
              if (currentCounter == this.counter) {
                  postBackCallback();
              }
          }, <number>options);
    }
}
```

```
<dot:TextBox Text="{value: SearchText}" UpdateTextAfterKeydown="true" Changed="{command: ShowSuggestions()}" PostBack.Deferred="delay" PostBack.DeferredOptions="500" />
```

The providers have to be registered on the client side:
```
dotvvm.deferredPostbackProviders["customConfirm"] = function() { 
    return new CustomConfirmPostBackDeferral(); 
}
```

The PostBack.Deferred and PostBack.DeferredOptions should render probably some data attributes or something like that - the postBack function should look whether the control has some postbacks deferred.

UPDATE:
The syntax on server might look like this as we may need to have multiple providers applied to the function. It would be nice to be able to combine multiple deferred postback providers (e.g. delay and then confirmation). The providers can have their own server-side implementation to be able to support multiple properties for each provider and to distinguish between events they are applied to.
```
<dot:TextBox ...>
    <PostBack.DeferredProviders>
        <dot:DelayPostBackProvider EventName="Click" Timeout="500" />
        <dot:ConfirmPostBackProvider EventName="Click" Message="Are you sure?" />
    </PostBack.DeferredProviders>
</dot:TextBox>
```

**Comment by exyi** [comment]
It should be possible to apply it only on some properties. For example KendoGrid has lot of command postback properties and only one of them will have to be confirmed by user.
I'd like it written in binding expression, like `{command: Confirm("Really?") && DeleteDatabase(DbName)}` or `{command: if Confirm(...) then Delete()}`

**Comment by tomasherceg** [comment]
I don't like such complicated expressions in the viewmodel commands. It wouldn't be able to apply those properties by styles. Also, the PostBack.DeferredProviders property could be inherited, so you could apply it e.g. on a page level.

**Comment by exyi** [comment]
It would be able to apply it by styles, although bit hacky ;). Nevertheless you can add a helper method.
I don't think you want to apply it globally very often because it can lead to quite unexpected behavior.

**Comment by tomasherceg** [comment]
Well, I can image quite simple global rule - take all buttons with Text="Remove" and add the confirmation. 
We really don't want to modify the command expressions using styles.

**Comment by exyi** [comment]
````
<dot:Button>
<PostBack.DeferredProviders>
    <dot:ConfirmPostBackProvider EventName="Click" Message="Are you sure?" />
</PostBack.DeferredProviders>
```

**Comment by tomasherceg** [comment]
I agree it's too long. We can help it a little e.g. by making the EventName property optional - if it is not specified, the provider would apply to all events. Or we can use PostBack.Handlers instead of DeferredProviders and dot:ConfirmHandler instead of dot:ConfirmPostBackProvider.

I really don't think it is a good idea to use the following syntax.

``` {command: Confirm("Really?") && DeleteDatabase(DbName)} ```

The only advantage of this is that it's short. However:
1) It is confusing, because Confirm is a client-side function and DeleteDatabase is a server-side one.
2) You cannot make all buttons on the page to use some postback deferral provider, at least not in a nice way.
3) What if the user types only {command: DeleteDatabase(DbName) && Confirm("Really?") }? We'd have to check for these situations and there is a lot of complexity in that.
4) I want the expressions in value and command bindings to be as simple as possible - allowing these things will lead to another spaghetti-code framework.

```
<dot:Button>
  <PostBack.Handlers>
    <dot:ConfirmHandler EventName="Click" Message="Are you sure?" />
  </PostBack.Handlers>
</dot:Button>
```

This is long, however IntelliSense will help you with that and it is stylable really easily. 
I also want the PostBack.Handlers property to be inherited to the child objects, so it would work also in a scenario when you use some third-party control but you need to add a confirmation there.

I'll invite the others to see this and tell us their opinion.

**Comment by tomasherceg** [comment]
Implemented in 187bb1d5, however there are several things which need to be done.

1. Write a documentation page and create a example of custom postback handler.
2. Modify the base control hierarchy in DotVVM to look like this: object > DotvvmBindableObject > DotvvmControl > DotvvmBindableControl > HtmlGenericControl > ....

**Comment by tomasherceg** [comment]
The control hierarchy change is prepared in the newtests branch.
The documentation page has been written.
